src/meta_memcache/base/base_cache_pool.py

```diff
@@ -0,0 +1,314 @@
+                cmd_flags.append(flag.value + bytes_value)
+        cmd.extend(sorted(cmd_flags))
+        return b" ".join(cmd) + ENDL
+
+    def _exec(
+        self,
+        command: MetaCommand,
+        key: Key,
+        value: Optional[bytes] = None,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Union[Miss, Value, Success, NotStored, Conflict]:
+        """
+        Gets a connection for the key and executes the command
+
+        You can override to implement retries, having
+        a fallback pool, etc.
+        """
+        try:
+            with self._get_pool(key).get_connection() as conn:
+                self._conn_send_cmd(
+                    conn,
+                    command=command,
+                    key=key,
+                    value=value,
+                    flags=flags,
+                    int_flags=int_flags,
+                    token_flags=token_flags,
+                )
+                return self._conn_recv_response(conn, flags=flags)
+        except MemcacheServerError:
+            if self._write_failure_tracker:
+                if command in (MetaCommand.META_DELETE, MetaCommand.META_SET):
+                    self._write_failure_tracker.add_key(key)
+            raise
+
+    def _exec_multi(
+        self,
+        command: MetaCommand,
+        keys: List[Key],
+        values: Optional[List[bytes]] = None,
+        flags: Optional[Set[Flag]] = None,
+        int_flags: Optional[Dict[IntFlag, int]] = None,
+        token_flags: Optional[Dict[TokenFlag, bytes]] = None,
+    ) -> Dict[Key, Union[Miss, Value, Success, NotStored, Conflict]]:
+        """
+        Groups keys by destination, gets a connection and executes the commands
+        """
+        if values and len(values) != len(keys):
+            raise ValueError("Values needs to match the number of keys")
+        key_value_map = {k: v for k, v in zip(keys, values or [])}
+        pool_key_map: DefaultDict[ConnectionPool, List[Key]] = defaultdict(list)
+        for key in keys:
+            pool_key_map[self._get_pool(key)].append(key)
+
+        results: Dict[Key, Union[Miss, Value, Success, NotStored, Conflict]] = {}
+        try:
+            for pool, pool_keys in pool_key_map.items():
+                with pool.get_connection() as conn:
+                    for key in pool_keys:
+                        self._conn_send_cmd(
+                            conn,
+                            command=command,
+                            key=key,
+                            value=key_value_map.get(key),
+                            flags=flags,
+                            int_flags=int_flags,
+                            token_flags=token_flags,
+                        )
+                    for key in pool_keys:
+                        results[key] = self._conn_recv_response(conn, flags=flags)
```

Comment:
How do we know the response is for the key? We're not passing the key so are we relying on the order?

Reply:
Yes, order is guaranteed. You can also request keys to be returned, but since it is still replied in the same order, I didn't think it was worth it (and key size is sometimes significant portion of the BW)